Gas cost optimisation

eth-r opened the pull request:
Currently the code causes a large number of unnecessary allocations because everything is handled as `bytes`. Replacing `bytes` with static `bytesX` where possible helps remove allocations and reduce code complexity. Other optimisations may follow.

eth-r committed:
Use static slices in easy places instead of allocating

Many functions in the library use the pattern `X.slice(Y, Z)` where Z is a
known constant <= 32. In these cases we can avoid allocations and simplify the
code by using the corresponding `bytesZ` type instead of `bytes`.

Provide `sliceZ()` functions for common values of Z and use them in obvious
locations. Also include some necessary utilities for using the static slices.

eth-r committed:
Use static slices for Merkle steps

eth-r committed:
Use static slices in `extractTarget`

The `uint24` is a bit different to reverse, otherwise similar to other static
slices.

eth-r committed:
Reuse the header allocation when validating header chain

eth-r committed:
Simplify header work validation

eth-r committed:
Remove unnecessary allocations in `p2wpkhFromPubkey()`

eth-r committed:
Remove allocations from merkle verification

In `verifyHash256Merkle()` we are consistently hashing pairs of `bytes32`. These
pairs fit in the 64-byte scratch space solidity provides. By using the scratch
space instead of abi encoding we can avoid a lot of allocations when verifying
merkle proofs.

eth-r committed:
Add ability to parse VarInt in-place

The code uses a lot of VarInt parsing, which previously relied on allocating new
`bytes` arrays for each VarInt. Changing this to read a VarInt from a larger
byte array enables further reductions in allocations.

eth-r committed:
Remove allocations from VarInt parsing

eth-r committed:
Optimise input and output extraction functions

Parsing VarInts at specified locations in a larger byte array lets us avoid
repeated slicing in loops.

eth-r committed:
Remove allocating loops from Vin and Vout validation

eth-r committed:
Provide "at" functions for extracting inputTxId and txIndex

lukasz-zimnoch commented on solidity/contracts/BTCUtils.sol:
```diff
@@ -214,9 +288,13 @@ library BTCUtils {
         uint256 _varIntDataLen;
         uint256 _scriptSigLen;
-        (_varIntDataLen, _scriptSigLen) = extractScriptSigLen(_input);
+        (_varIntDataLen, _scriptSigLen) = extractScriptSigLen(_input, _at);
```
Should be `extractScriptSigLenAt` instead of `extractScriptSigLen`.

tomaszslabon commented on solidity/contracts/BTCUtils.sol:
```diff
@@ -27,13 +27,17 @@ library BTCUtils {
+        return determineVarIntDataLengthAt(_flag, 0);
+    }
+
+    function determineVarIntDataLengthAt(bytes memory _b, uint256 _at) internal pure returns (uint8) {
```
Could we add a brief description to this function (may be similar to the `determineVarIntDataLength`)?

tomaszslabon commented on solidity/contracts/BTCUtils.sol:
```diff
@@ -46,15 +50,26 @@ library BTCUtils {
+        return parseVarIntAt(_b, 0);
+    }
+
+    function parseVarIntAt(bytes memory _b, uint256 _at) internal pure returns (uint256, uint256) {
```
Could we add a brief description to this function?

tomaszslabon commented on solidity/contracts/BTCUtils.sol:
```diff
@@ -109,6 +156,16 @@ library BTCUtils {
         return _number;
     }
 
+    function bytesToUint(bytes4 _b) internal pure returns (uint256) {
```
Could we add a bries description here?

tomaszslabon commented on solidity/contracts/BTCUtils.sol:
```diff
@@ -293,15 +371,23 @@ library BTCUtils {
+        return _input.slice32(0);
+    }
+
+    function extractInputTxIdLeAt(bytes memory _input, uint256 _at) internal pure returns (bytes32) {
```
Could we add a brief description?

tomaszslabon commented on solidity/contracts/BTCUtils.sol:
```diff
@@ -599,12 +683,17 @@ library BTCUtils {
         return hash256(abi.encodePacked(_a, _b));
     }
 
+    function _hash256MerkleStep(bytes32 _a, bytes32 _b) internal view returns (bytes32) {
```
Could we add a brief description?

tomaszslabon commented on solidity/contracts/BytesLib.sol:
```diff
@@ -270,6 +270,62 @@ library BytesLib {
+    }
+
+    // Static slice functions, no bounds checking
+    function slice32(bytes memory _bytes, uint _start) internal pure returns (bytes32 res) {
+        assembly {
+            res := mload(add(add(_bytes, 32), _start))
+        }
+    }
+
+    function slice20(bytes memory _bytes, uint _start) internal pure returns (bytes20) {
+        return bytes20(slice32(_bytes, _start));
+    }
+
+    function slice8(bytes memory _bytes, uint _start) internal pure returns (bytes8) {
+        return bytes8(slice32(_bytes, _start));
+    }
+
+    function slice4(bytes memory _bytes, uint _start) internal pure returns (bytes4) {
+        return bytes4(slice32(_bytes, _start));
+    }
+
+    function slice3(bytes memory _bytes, uint _start) internal pure returns (bytes3) {
+        return bytes3(slice32(_bytes, _start));
+    }
+
+    function slice2(bytes memory _bytes, uint _start) internal pure returns (bytes2) {
+        return bytes2(slice32(_bytes, _start));
+    }
+
```
Could we add brief descriptions to these functions?

eth-r committed:
Fix typo in determineInputLengthAt

Should call `extractScriptSigLenAt`, not `extractScriptSigLen`

eth-r committed:
Provide basic documentation in BytesLib new functions

eth-r committed:
Document new functions in BTCUtils and remove unused one

eth-r committed:
Use scratch space in `hash256View`

The first 64 bytes of memory are available as scratch space within inline
assembly. Using the constant memory location 0x00 instead of loading it from the
unallocated area as a variable should reduce gas cost marginally.

eth-r committed:
Improve merkle proof efficiency

Improve efficiency of `hash256View()` and use it over `hash256()` wherever
possible. Remove intermediate allocations from `ValidateSPV.prove()`.

eth-r committed:
Use static sizes for `ValidateSPV.calculateTxId()` arguments

The version and locktime are always 4 bytes, so use `bytes4` instead of `bytes`.
